### PR TITLE
Added /usersearch and /useractivity methods

### PR DESCRIPTION
**Pull request by geoffdutton**
Wrappers around these:

https://amplitude.zendesk.com/hc/en-us/articles/205469748-Dashboard-Rest-API-Export-Amplitude-Dashboard-Data#user%20search

https://amplitude.zendesk.com/hc/en-us/articles/205469748-Dashboard-Rest-API-Export-Amplitude-Dashboard-Data#user-activity

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 node_modules/
+.idea
```
Comment (crookedneighbor) [inline .gitignore]: Please remove this.

File: README.md [hunk 2]
```diff
@@ -143,6 +143,39 @@ amplitude.export({
...
+```javascript
+var amplitude = new Amplitude('api-token', { secretKey: 'secret' })
+
+amplitude.userSearch('some-id-or-user-prop', {withUserActivity:: true})
+```
+
+
+## User Activity
+
+The export method requires your [secret key](https://amplitude.zendesk.com/hc/en-us/articles/206728448-Where-can-I-find-my-app-s-API-Key-or-Secret-Key-) to be added when initializing the amplitude object. This method uses the [dashboard api](https://amplitude.zendesk.com/hc/en-us/articles/205469748-Dashboard-Rest-API-Export-Amplitude-Dashboard-Data#user-activity).
```
Comment (crookedneighbor) [inline README.md]: export method -> user activity method

File: amplitude.js [hunk 3]
```diff
@@ -84,12 +87,69 @@ Amplitude.prototype.export = function (options) {
...
+Amplitude.prototype.userActivity = function (amplitudeId, opts) {
+  opts = opts || {}
+  opts.user = amplitudeId
+
+  if (!this.secretKey) {
+    throw new Error('secretKey must be set to use the export method')
+  }
+
+  if(!amplitudeId) {
+    throw new Error('amplitude_id for must be passed.')
```
Comment (crookedneighbor) [inline amplitude.js]: This sentence does not make grammatic sense.

File: amplitude.js [hunk 4]
```diff
@@ -84,12 +87,69 @@ Amplitude.prototype.export = function (options) {
...
+  })
+
+}
+
+Amplitude.prototype.userActivity = function (amplitudeId, opts) {
+  opts = opts || {}
+  opts.user = amplitudeId
+
+  if (!this.secretKey) {
+    throw new Error('secretKey must be set to use the export method')
```
Comment (crookedneighbor) [inline amplitude.js]: Export method referenced here.

File: amplitude.js [hunk 5]
```diff
@@ -84,12 +87,69 @@ Amplitude.prototype.export = function (options) {
...
     .query({
       start: options.start,
       end: options.end
     })
 }
 
+Amplitude.prototype.userSearch = function (userSearchId, opts) {
+
+  if (!this.secretKey) {
+    throw new Error('secretKey must be set to use the export method')
```
Comment (crookedneighbor) [inline amplitude.js]: Export method.

File: test/userSearch.test.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+      console.error(err)
+      expect(err).to.not.exist
+    })
+  })
+
+  it('resolves matches found by a user property', function () {
+    let search = 'user_id_2'
+    let mockedRequest = generateMockedRequest(search, this.userSearchIds.found_by_user_props, 200)
+
+
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Why so many blank spaces?

File: test/userSearch.test.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+describe('userSearch', function () {
+  let userActivityStub
+
+  beforeEach(function () {
+    this.amplitude = new Amplitude('token', {
+      secretKey: 'key'
+    })
+
+    userActivityStub = sinon.stub(this.amplitude, 'userActivity')
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Let's use `this.userActivityStub` instead

File: test/userSearch.test.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  it('throws an error if secret key is missing', function () {
+    delete this.amplitude.secretKey
+
+    expect(() => {
+      this.amplitude.userSearch('anything')
+    }).to.throw('secretKey must be set to use the export method')
+  });
+
+  it('throws an error if nothing passed', function () {
+
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Extra line break here.

File: test/userSearch.test.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+
+    return this.amplitude.userSearch(search).then((res) => {
+      expect(res.matches).to.be.a('array')
+      expect(res.matches.length).to.eql(1)
+      expect(res.type).to.eql('match_user_props')
+      mockedRequest.done()
+    }).catch((err) => {
+      console.error(err)
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Remove this.

File: test/userSearch.test.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    let mockedRequest = generateMockedRequest(search, this.userSearchIds.found_by_amplitude_id, 200)
+
+    return this.amplitude.userSearch(search).then((res) => {
+      expect(res.matches).to.be.a('array')
+      expect(res.matches.length).to.eql(1)
+      expect(res.type).to.eql('match_amplitude_id')
+      expect(userActivityStub.calledOnce).to.be.false
+      mockedRequest.done()
+    }).catch((err) => {
+      console.error(err)
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Remove this.

File: test/userSearch.test.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    let search = 'user_id_3'
+    let mockedRequest = generateMockedRequest(search, this.userSearchIds.found_by_user_or_device_id, 200)
+
+    return this.amplitude.userSearch(search).then((res) => {
+      expect(res.matches).to.be.a('array')
+      expect(res.matches.length).to.eql(1)
+      expect(res.type).to.eql('match_user_or_device_id')
+      mockedRequest.done()
+    }).catch((err) => {
+      console.error(err)
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Remove.

File: test/userSearch.test.js [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    let search = 'cant-find-me'
+    let mockedRequest = generateMockedRequest(search, this.userSearchIds.not_found, 200)
+
+    return this.amplitude.userSearch(search).then((res) => {
+      expect(res.matches).to.be.a('array')
+      expect(res.matches.length).to.eql(0)
+      expect(res.type).to.eql('nomatch')
+      mockedRequest.done()
+    }).catch((err) => {
+      console.error(err)
```
Comment (crookedneighbor) [inline test/userSearch.test.js]: Remove.

File: amplitude.js [hunk 13]
```diff
@@ -84,12 +87,69 @@ Amplitude.prototype.export = function (options) {
...
+  return request.get(amplitudeSecretEndPoint + '/api/2/export')
     .auth(this.token, this.secretKey)
     .query({
       start: options.start,
       end: options.end
     })
 }
 
+Amplitude.prototype.userSearch = function (userSearchId, opts) {
+
```
Comment (crookedneighbor) [inline amplitude.js]: Extra line break.

File: amplitude.js [hunk 14]
```diff
@@ -84,12 +87,69 @@ Amplitude.prototype.export = function (options) {
...
+    throw new Error('value to search for must be passed.')
+  }
+
+  opts = opts || {}
+
+  var userActivity = this.userActivity.bind(this)
+
+  return request.get(amplitudeSecretEndPoint + '/api/2/usersearch')
+  .auth(this.token, this.secretKey)
+  .query({
```
Comment (crookedneighbor) [inline amplitude.js]: This should be indebted to match the style of the other methods.

File: amplitude.js [hunk 15]
```diff
@@ -84,12 +87,69 @@ Amplitude.prototype.export = function (options) {
...
+    user: userSearchId
+  })
+  .set('Accept', 'application/json')
+  .then(function (res) {
+
```
Comment (crookedneighbor) [inline amplitude.js]: Extra line break.

File: README.md [hunk 16]
```diff
@@ -143,6 +143,39 @@ amplitude.export({
...
+
+Get a user summary and their recent events. The method returns a 'userData' object. This method requires an amplitude_id. You can use userSearch method to find that.
+
+```javascript
+var amplitude = new Amplitude('api-token', { secretKey: 'secret' })
+
+amplitude.userActivity('amplitude_id')
```
Comment (crookedneighbor) [inline README.md:200]: Let's include how it resolves and provide something some examples for the properties returned.

File: README.md [hunk 17]
```diff
@@ -143,6 +143,39 @@ amplitude.export({
...
+## User Search
+
+The export method requires your [secret key](https://amplitude.zendesk.com/hc/en-us/articles/206728448-Where-can-I-find-my-app-s-API-Key-or-Secret-Key-) to be added when initializing the amplitude object. This method uses the [dashboard api](https://amplitude.zendesk.com/hc/en-us/articles/205469748-Dashboard-Rest-API-Export-Amplitude-Dashboard-Data#user%20search).
+
+Search for a user with a specified Amplitude ID, Device ID, User ID, or User ID prefix. The method returns a 'matches' object.
+
+```javascript
+var amplitude = new Amplitude('api-token', { secretKey: 'secret' })
+
+amplitude.userSearch('some-id-or-user-prop')
```
Comment (crookedneighbor) [inline README.md:155]: Let's include how it resolves and provide something some examples for the properties returned.

**Review by crookedneighbor — CHANGES_REQUESTED** [review]
Left some comments, but looks mostly good to me.

File: .gitignore (hunk 1 above)
Comment (geoffdutton) [reply .gitignore]: Do you want the .idea folder to be committed in the repo? It's the folder generated by the Intellij IDEA IDE.

**Comment by geoffdutton** [comment]
I've addressed your comments. Thanks for reviewing.

File: .gitignore (hunk 1 above)
Comment (crookedneighbor) [reply .gitignore]: No. you should have the .idea folder as an part of your global gitignore. Not part of this repo.

**Comment by geoffdutton** [comment]
Ah, got it.

File: README.md (hunk 17 above)
Comment (geoffdutton) [reply README.md:155]: Are the examples I added what you were looking for?

**Comment by crookedneighbor** [comment]
Made a few changes and took out the `withUserActivity` shortcut in the `userSearch` method, but added an example in the readme for how to do it yourself. I think it might make more sense to have a separate method that calls both methods internally, but I want to sit on it a bit before committing to that being a part of the modules API.

How do you want to be credited? I need your preferred name, email and website url. If I don't hear from you in the next few days, I'll just use what's on your git commit.

``` json
{
  "name": "Geoff Dutton",
  "email": "g.dutton@gmail.com"
}
```

**Comment by geoffdutton** [comment]
Sorry, I missed this. That's fine for credit, thank you.

And yes, I was going back and fourth on the `withUserActivity`. It's not part of the amplitude api spec, the main reason I added it is because I have a bunch of spots in some of my production apps that I have been using it.

I'm fine with leaving it out though.

**Comment by crookedneighbor** [comment]
Sorry for the delay on this. The latest release has your changes. Thanks for the work!

**Pull request closed without merging**